MockerTests/DynamicMocksTests.swift

```diff
@@ -0,0 +1,294 @@
+                                         dataType: .json,
+                                         statusCode: 200,
+                                         data: [.get: goodMockData])
+        // succeedingMock.register() - Do NOT call now as will replace failingMock.
+        // Instead, return `succeedingMock` using new `onMockFor` handler if problem is fixed:
+
+        let refreshTokenMock = Mock(url: FakeService.FakeServiceURLs.refreshTokenURL,
+                                    dataType: .json,
+                                    statusCode: 200,
+                                    data: [.get: goodTokenData])
+        refreshTokenMock.register()
+
+        // Now set up our dynamic mock that will replace the failingProfileMock
+        // with the succeedingProfileMock IFF we have a valid token (if the refreshToken call is made
+        // and gets a valid token)
+
+        Mocker.onMockFor = { _ in
+            if FakeService.accessToken.token == "good token" {
+                return succeedingProfileMock
+            } else { return nil }
+        }
+
+        let expectFailure = expectation(description: "call url - should get error")
+
+        // Now make the profile call that we expect will fail, internally
+        // get a new token from the service, and then retry and finally return us valid profile data.
+        service.profile(for: "JJ") { profile, error in
+            XCTAssertNil(error)
+            XCTAssertNotNil(profile)
+            XCTAssertEqual(profile!.name, "Joe Jones")
+            expectFailure.fulfill()
+        }
+
+        wait(for: [expectFailure], timeout: 2.0)
+    }
+
+    func FAILS_testOnMockForConditionalReplacementAvdLeeSuggestion1() {
+        let service = FakeService() // the service we are testing
+
+        // Create Mock data
+        let goodMockData = "{\"name\":\"Joe Jones\"}".data(using: .utf8)!
+        let goodTokenData = "{\"token\":\"good token\"}".data(using: .utf8)!
+
+        // Create our three Mocks:
+
+        let succeedingProfileMock = Mock(url: FakeService.FakeServiceURLs.profileURL,
+                                         dataType: .json,
+                                         statusCode: 200,
+                                         data: [.get: goodMockData])
+        // succeedingMock.register() - Do NOT call now as will get replaced by failingMock.
+        // Instead, try registering it in the `onRequest` handler for failingProfileMock
+        // if the token has been updated
+
+        var failingProfileMock = Mock(url: FakeService.FakeServiceURLs.profileURL,
+                                      dataType: .json,
+                                      statusCode: 401,
+                                      data: [.get: Data()])
+
+        failingProfileMock.onRequest = { _, _ in
+            // this won't help because it's too late.
+            if FakeService.accessToken.token == "good token" {
```

Comment:
I've checked out your branch and commented out this code. Then the test succeeded.

You have to register the succeeding mock as soon as it failed once.

Reply:
Hi @AvdLee - thanks for taking a look at it.

Yes.  Unfortunately, commenting out that line removes testing steps 1 & 2 of the four aspects of the refresh portion of the code under test that I listed above.  That's a problem because if the dynamic token refresh handling code fails to process the mocked refresh token payload properly and deliver the results to the right place(s), then the test will still pass. 

It is not possible to test this in Mocker without some way to dynamically replace a mock when Mocker is looking up which mock to use because there's no place else to test that the chained network call did it's work properly (since it's chained control doesn't return to the test code itself during the chained processing).

This simple example code for this sample test doesn't make it look like there's much to do, but in the real world it's more complicated and something that really wants to be tested (refreshing the accessToken with a refreshToken can fail because the refreshToken has expired; then you have to get a new refreshToken, which then allows you to get a new accessToken, etc).

This is a specific instance of the general limitation which the `onMockFor` addition addresses: there is currently no way to dynamically set a mock after processing the payload of a mocked network call within a series of chained networking calls when two of the calls call the same url (well, possible if there are two or more network calls between the repeated call to the same URL, but not if is only 1).

Reply:
This is a specific instance of the general limitation which the `onMockFor` addition addresses: there is currently no way to dynamically set a mock after processing the payload of a mocked network call within a series of chained networking calls when two of the calls call the same url (well, possible if there are two or more network calls between the repeated call to the same URL, but not if is only 1).

Imagine testing a multi-step combine chain which makes repeated calls to a network endpoint and you want to change the mock for each time you call the endpoint but you want/need to test the results of the combine chain operators between each call to the endpoint to figure out which mock to activate - that's another example of the use case for `onMockFor`.

Reply:
Correct me if I'm wrong but don't you want to simply verify that the correct new token is sent with the second succeeding request?

If so, you can do this with the `onRequest` of the succeeding mock and assert in there

Reply:
Interesting idea!  In this simplified example test situation that is indeed all that needs to be tested. I modified the test and it works as you suggest.  Nice!

I'll have to go dig into my real-world shipping-code use-case and see if something similar can work there.  It's a quite a bit more complicated; I have variants on the tests that test the error handling paths via mocks that simulate the intermediate service requests returning various errors and making sure they either bubble up through the combine pipeline to the original caller properly or trigger alternate auto-try-to-fix-and-retry handlers and then retry again.

Thanks for the feedback.

Reply:
Tried it out in the real app tests and this approach does work for the sub-set of tests I tried it on there also (there are more, but some will take more work to rework).

While I do still think the `onMockfor` functionality will be useful in other situations, it's clearly not necessary for this one, and the no-changes-needed approach always wins :)

Thank you for your patience and your assistance helping me see an alternative no-changes-needed approach.

Reply:
All good, glad I could help! 💪 Thanks for staying motivated to integrate my feedback!